WorkloadManagementSystem/Utilities/QueueUtilities.py

```diff
@@ -0,0 +1,225 @@
+"""Utilities to help Computing Element Queues manipulation
+"""
+
+__RCSID__ = '$Id$'
+
+from DIRAC import S_OK, S_ERROR
+from DIRAC.Core.Utilities.List import fromChar
+from DIRAC.Core.Utilities.ClassAd.ClassAdLight import ClassAd
+from DIRAC.ConfigurationSystem.Client.Helpers.Resources import getDIRACPlatform
+from DIRAC.ConfigurationSystem.Client.Helpers.Operations import Operations
+from DIRAC.Core.Security.ProxyInfo import getProxyInfo
+
+
+def getQueuesResolved(siteDict):
+  """
+  Get the list of queue descriptions merging site/ce/queue parameters and adding some
+  derived parameters.
+
+  :param dict siteDict: dictionary with configuration data as returned by Resources.getQueues() method
+
+  :return: S_OK/S_ERROR, Value dictionary per queue with configuration data updated, e.g. for SiteDirector
+  """
+
+  queueDict = {}
+
+  for site in siteDict:
+    for ce in siteDict[site]:
+      ceDict = siteDict[site][ce]
+      qDict = ceDict.pop('Queues')
+      for queue in qDict:
+        queueName = '%s_%s' % (ce, queue)
+        queueDict[queueName] = qDict[queue]
+        queueDict[queueName] = qDict[queue]
+        queueDict[queueName]['Queue'] = queue
+        queueDict[queueName]['Site'] = site
+        # Evaluate the CPU limit of the queue according to the Glue convention
+        # To Do: should be a utility
+        if "maxCPUTime" in queueDict[queueName] and \
+           "SI00" in queueDict[queueName]:
+          maxCPUTime = float(queueDict[queueName]['maxCPUTime'])
+          # For some sites there are crazy values in the CS
+          maxCPUTime = max(maxCPUTime, 0)
+          maxCPUTime = min(maxCPUTime, 86400 * 12.5)
+          si00 = float(queueDict[queueName]['SI00'])
+          queueCPUTime = 60. / 250. * maxCPUTime * si00
+          queueDict[queueName]['CPUTime'] = int(queueCPUTime)
+
+        # Tags & RequiredTags defined on the Queue level and on the CE level are concatenated
+        # This also converts them from a string to a list if required.
+        for tagFieldName in ('Tag', 'RequiredTag'):
+          ceTags = ceDict.get(tagFieldName, [])
+          if isinstance(ceTags, basestring):
+            ceTags = fromChar(ceTags)
+          queueTags = queueDict[queueName].get(tagFieldName)
+          if queueTags and isinstance(queueTags, basestring):
+            queueTags = fromChar(queueTags)
+            queueDict[queueName][tagFieldName] = queueTags
+          if ceTags:
+            if queueTags:
+              allTags = list(set(ceTags + queueTags))
+              queueDict[queueName][tagFieldName] = allTags
+            else:
+              queueDict[queueName][tagFieldName] = ceTags
+
+        # Some parameters can be defined on the CE level and are inherited by all Queues
+        for parameter in ['MaxRAM', 'NumberOfProcessors', 'WholeNode']:
```

Comment:
I think this would be clearer if written as:
```suggestion
        for parameter in ['MaxRAM', 'NumberOfProcessors', 'WholeNode']:
          queueParameter = queueDict[queueName].get(parameter, ceDict.get(parameter))
          if queueParameter:
            queueDict[queueName][parameter] = queueParameter
```